app/src/main/java/org/dhis2/usescases/main/program/ProgramPresenter.kt

```diff
@@ -1,55 +1,33 @@
 package org.dhis2.usescases.main.program
 
-import android.content.Context
-import android.os.Bundle
-import android.text.TextUtils.isEmpty
 import io.reactivex.disposables.CompositeDisposable
-import io.reactivex.functions.Action
-import io.reactivex.functions.Consumer
 import io.reactivex.processors.PublishProcessor
 import org.dhis2.data.prefs.PreferenceProvider
 import org.dhis2.data.schedulers.SchedulerProvider
 import org.dhis2.data.tuples.Pair
-import org.dhis2.usescases.datasets.datasetDetail.DataSetDetailActivity
-import org.dhis2.usescases.programEventDetail.ProgramEventDetailActivity
-import org.dhis2.usescases.searchTrackEntity.SearchTEActivity
-import org.dhis2.utils.ColorUtils
-import org.dhis2.utils.Constants
-import org.dhis2.utils.analytics.SELECT_PROGRAM
-import org.dhis2.utils.analytics.TYPE_PROGRAM_SELECTED
+import org.dhis2.utils.Constants.PROGRAM_THEME
 import org.dhis2.utils.filters.FilterManager
-import org.dhis2.utils.granularsync.GranularSyncContracts
-import org.dhis2.utils.granularsync.SyncStatusDialog
 import org.hisp.dhis.android.core.period.DatePeriod
-import org.hisp.dhis.android.core.program.ProgramType
 import timber.log.Timber
 
-/**
- * Created by ppajuelo on 18/10/2017.f
- */
-
 class ProgramPresenter internal constructor(
+    private val view: ProgramView,
     private val homeRepository: HomeRepository,
     private val schedulerProvider: SchedulerProvider,
-    private val preferenceProvider: PreferenceProvider
-) : ProgramContract.Presenter {
+    private val preferenceProvider: PreferenceProvider,
+    private val filterManager: FilterManager
+) {
 
-    private var view: ProgramContract.View? = null
-    private var compositeDisposable: CompositeDisposable? = null
+    var disposable: CompositeDisposable = CompositeDisposable()
     private val programQueries = PublishProcessor.create<Pair<List<DatePeriod>, List<String>>>()
 
-    override fun init(view: ProgramContract.View) {
-        this.view = view
-        this.compositeDisposable = CompositeDisposable()
-
-        val loadingProcessor = PublishProcessor.create<Boolean>()
-
-        compositeDisposable!!.add(
-            FilterManager.getInstance().asFlowable()
-                .startWith(FilterManager.getInstance())
+    fun init() {
+        disposable.add(
+            filterManager.asFlowable()
+                .startWith(filterManager)
                 .doOnNext { Timber.tag("INIT DATA").d("NEW FILTER") }
                 .switchMap { filterManager ->
-                    loadingProcessor.onNext(true)
+                    view.showFilterProgress()
```

Comment:
I thought I changed this as it can cause a crash as we are accessing a ui view from an io thread

Reply:
true, this is emitting in IO scheduler

Reply:
done